clustermesh: add GlobalServiceCache in clustermesh/common package

MrFreezeex opened the pull request:
Please ensure your pull request adheres to the following guidelines:

- [ ] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [ ] If your commit description contains a `Fixes: <commit-id>` tag, then
      please add the commit author[s] as reviewer[s] to this issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [ ] Provide a title or release-note blurb suitable for the release notes.
- [ ] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)
- [ ] Thanks for contributing!

With the ongoing work on EndpointSlice synchronization, we need to have a `globalServiceCache` there as well so this commit is refactoring the GlobalServiceCache to be inside the common clustermesh package so that we can avoid code duplication and access to this struct outside of the pkg/clustermesh package.

It also refactor the GlobalServiceCache in a separate to use `types.NamespacedName` instead of a simple string.

```release-note
Refactor clustermesh global service cache to prepare for the endpoint slice clustermesh synchronization
```

MrFreezeex committed:
clustermesh: add GlobalServiceCache in clustermesh/common package

With the ongoing work on EndpointSlice synchronization, we need to have
a `globalServiceCache` there as well so this commit is refactoring
the GlobalServiceCache to be inside the common clustermesh package so that
we can avoid code duplication and access to this struct outside of the
pkg/clustermesh package.

Signed-off-by: Arthur Outhenin-Chalandre <arthur@cri.epita.fr>

MrFreezeex committed:
clustermesh: refactor GlobalServiceCache to use types.NamespacedName

Refactor globalServiceCache byName map to use types.NamespacedName
instead of string where the name and namespace was concatenated with a
`/` as a separator. Using this type makes the existing behavior more
explicit/type safe and prevent potential error where we would pass a
potential malformed string.

Signed-off-by: Arthur Outhenin-Chalandre <arthur@cri.epita.fr>

giorio94 commented on pkg/service/store/store.go:
```diff
@@ -89,6 +91,11 @@ func (s *ClusterService) NamespaceServiceName() string {
+// ServiceNamespacedName returns the namespace and service name with types.NamespacedName
+func (s *ClusterService) ServiceNamespacedName() types.NamespacedName {
+	return types.NamespacedName{Name: s.Name, Namespace: s.Namespace}
+}
```
Sorry, I just realized that I never published this comment on the other PR:

> I'd personally avoid to have both NamespaceServiceName() and ServiceNamespacedName(), as possibly confusing. What about modifying NamespaceServiceName() to return a NamespacedName, and convert possible remaining calling sites (there shouldn't be any other except from the ones you already modified) over to NamespaceServiceName().String() if a string is needed?

MrFreezeex replied:
Ah yesss, I checked and it was the only place used it seems... I supposed it was used in many places but not even :sweat_smile:.

giorio94 commented on pkg/clustermesh/common/services.go:
```diff
@@ -136,7 +136,7 @@ func (c *GlobalServiceCache) OnDelete(svc *serviceStore.ClusterService) bool {
 
 	if GlobalService, ok := c.byName[svc.NamespaceServiceName()]; ok {
-		return c.delete(GlobalService, svc.Cluster, types.NamespacedName{Name: svc.Name, Namespace: svc.Namespace}.String())
+		return c.delete(GlobalService, svc.Cluster, types.NamespacedName{Name: svc.Name, Namespace: svc.Namespace})
```
Nit. Could use `svc.NamespaceServiceName` instead of constructing the NamespacedName.

giorio94 commented on pkg/clustermesh/common/services.go:
```diff
@@ -0,0 +1,163 @@
+		newGlobalSvc := newGlobalService()
+		for k, v := range globalSvc.ClusterServices {
+			newGlobalSvc.ClusterServices[k] = v
+		}
```
Nit. Could use `maps.Clone()` instead of the bespoke logic.

giorio94 commented on pkg/clustermesh/common/services.go:
```diff
@@ -0,0 +1,163 @@
+	return false
+}
+
+// GetGlobalService returns the service for a specific cluster
```
Nit. The comment should be updated, as the method returns the global service. I'd also mention that the underlying ClusterServices must not be mutated, as not deep copied.

giorio94 commented on pkg/clustermesh/common/services.go:
```diff
@@ -0,0 +1,163 @@
+	// global service
+	if len(globalService.ClusterServices) == 0 {
+		scopedLog.Debugf("Deleted global service %s", key)
+		globalSvc := c.byName[key]
```
Nit. You could use `globalService` instead of retrieving it again.

giorio94 commented on pkg/clustermesh/common/services.go:
```diff
@@ -0,0 +1,163 @@
+	if len(globalService.ClusterServices) == 0 {
+		scopedLog.Debugf("Deleted global service %s", key)
+		globalSvc := c.byName[key]
+		c.byNamespace[key].Delete(globalSvc)
```
The key should be the namespace only, and not the combination of namespace and name. I know that it is addressed in the subsequent commit, but ideally every commit should be correct to simplify bisecting/cherry picking. One possibility could be extracting the introduction of the `byNamespace` concept in a third commit on top of the others, which would also make it easier to discern the no-op copy and paste of the existing methods from the subsequent additions.

MrFreezeex replied:
Ohh yes indeed, I ended up putting those in a separate commit after the string -> types.NamespaceName change

giorio94 reviewed (commented):
Thanks! lgtm pending a few minor comments inline.

MrFreezeex committed:
clustermesh: add byNamespace map to the global service cache

Also store service in the GlobalServiceCache by namespace so that we can
get all the services inside a namespace which is needed by the
EndpointSlice synchronization feature.

Signed-off-by: Arthur Outhenin-Chalandre <arthur@cri.epita.fr>

giorio94 reviewed (approved):
Thanks!
